core/controller/src/main/scala/whisk/core/controller/actions/PrimitiveActions.scala

```diff
@@ -139,6 +199,356 @@ protected[actions] trait PrimitiveActions {
+                  entitlementProvider.check(user, Privilege.ACTIVATE, Set(resource), noThrottle = true).onComplete {
+                    case Failure(t) =>
+                      // failed entitlement check
+                      val response =
+                        ActivationResponse.applicationError(compositionComponentNotAccessible(next.toString))
+                      completeAppActivation(user, session, response)
+                    case Success(_) =>
+                      // successful entitlement check
+                      WhiskActionMetaData.resolveActionAndMergeParameters(entityStore, fqn).onComplete {
+                        case Failure(t) =>
+                          // resolution failure
+                          val response =
+                            ActivationResponse.applicationError(compositionComponentNotFound(next.toString))
+                          completeAppActivation(user, session, response)
+                        case Success(next) =>
+                          if (session.accounting.components >= actionSequenceLimit) {
+                            // composition is too long
+                            val response = ActivationResponse.applicationError(compositionIsTooLong)
+                            completeAppActivation(user, session, response)
+                          } else {
+                            // successful resolution
+                            invokeComponent(user, next, params, session)
+                          }
+                      }
+                  }
+              }
+          }
+      }
+    }
+  }
+
+  /**
+   * A method that knows how to handle a component invocation.
+   *
+   * This method distinguishes primitive actions, sequences, and compositions.
+   * If the component action is a composition, it invokes invokeComposition and returns.
+   * invokeComposition takes care of resuming the execution of this composition, when the nested composition finishes.
+   * If the component action is not a composition, it is invoked followed by the reinvocation of the conductor action.
+   * This method also keeps track of the duration and memory footprint for the composition.
+   *
+   * @param user the identity invoking the action
+   * @param action the component action to invoke
+   * @param payload the dynamic arguments for the activation
+   * @param session the session object for this composition
+   * @param transid a transaction id for logging
+   */
+  private def invokeComponent(user: Identity, action: WhiskActionMetaData, payload: Option[JsObject], session: Session)(
+    implicit transid: TransactionId) {
+
+    val exec = action.toExecutableWhiskAction
+    if (action.annotations.get("conductor").isDefined && exec.isDefined) {
+      // composition
+      invokeComposition(user, exec.head, payload, None, Some(session.activationId), Some(session)) // non-blocking
```

Comment:
s/head/get

Reply:
maybe this is clearer and avoids the gets.
```
action.toExecutableWhiskAction match {
   case Some(x) if !conductor => invoke simple action
   case Some(x) => invoke conductor
   case None => invoke sequence
}
```

Reply:
🙏 for clarity, use argument names in the method calls, particularly when using None - makes it easier to remember what value that is for.

Reply:
have to go back and look, is the conductor accounted for in invoke composition?
edit: yes it is, perhaps a comment, since the component accounting is increased below - for clarity.

Reply:
OK for match.